_posts/2020-09-25-programmatic-authentication-under-iap.md

```diff
@@ -0,0 +1,200 @@
+proxy.addHeader(port, headersToSet, function (err, resp) {
+   if (err) {
+       console.error("Error encountered", err);
+   } else {
+       console.info(`Headers added ${JSON.stringify(headersToSet)}, ${resp}`);
+   }
+});
+```
+
+This allowed us to add the authorization bearer token to every request that was being made through the Protractor browser instance.
+Additionally, we also had to set the `trustAllServers` parameter in browsermob-proxy to `true` to disable verification of all upstream servers' SSL certificates.
+
+The addition of the proxy option to protractor browser capabilities was done to have Protractor route all it’s browser requests through the proxy spawned at the designated `host: port` combination.
+
+```javascript
+proxy: {
+       'proxyType': 'manual',
+       'httpProxy': `${this.params.proxyHost}:${this.params.proxyPort}`,
+       'sslProxy': `${this.params.proxyHost}:${this.params.proxyPort}`
+       }
+```
+
+In our node-based automations, where we were making use of the request module to make API calls for their assertion.
+
+```javascript
+exports.customRequest = function (options, callback) {
+   var base;
+   if (browser.params.isRunOnTestApp) {
+       base = request.defaults({
+           headers: {
+               'Authorization': iapAuthToken
+           },
+           jar: true
+       });
+   }
+   else {
+       base = request.defaults({
+           jar: true
+       });
+   }
+   base(options, callback);
+};
+```
+
+We overrode them and to accommodate the addition of the authorization bearer token in every request that was being made.
+
+And for all our load tests that were being run through JMeter, we made use of a similar script as mentioned in the first example above. It generates the access token and writes it to a file. This file is then read by the JMX scripts while adding the authorization bearer token to all the network requests that are made through JMeter.
+Keeping in mind that these tokens have a short expiration, we generate the token and write them to the files as and when the load tests are run.
+
+### Aftermath
+
+The IAP is definitely a good option to regulate access to private resources without having to implement a VPN tunneling. But throwing accessibility to automated agents in the mix created new interesting challenges here. There are some obvious pros to it:
+
+1. Easy access regulation to resources
+2. Fine grained control of access through users' and services' IAM roles
+3. Access through untrusted networks without the requirements of a VPN tunnel
+
+But there are some cons that we discovered along the way of implementing programmatic authentication:
+
+1. Slower execution speeds on private resources, a resultant of re-routing the access which is done to add authorization headers
+2. Added overhead of handling programmatic authentication in automation tests, which does not cater to the user flow being tested. Users won't be accessing the actual app over IAP.
```

Comment:
usage of programmatic is redundant when talking about automation

Reply:
removed 'programmatic'